### PR TITLE
Add design for repository maintenance job

### PR DESCRIPTION
**Pull request by qiuming-best**
Thank you for contributing to Velero!

# Please add a summary of your change

# Does your change fix a particular issue?

Fixes #(issue)
https://github.com/vmware-tanzu/velero/issues/7291
# Please indicate you've done the following:

- [ ] [Accepted the DCO](https://velero.io/docs/v1.5/code-standards/#dco-sign-off). Commits without the DCO will delay acceptance.
- [ ] [Created a changelog file](https://velero.io/docs/v1.5/code-standards/#adding-a-changelog) or added `/kind changelog-not-required` as a comment on this pull request.
- [ ] Updated the corresponding documentation in `site/content/docs/main`.

### CODE REVIEW

File: design/repository-maintenance.md [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+			return containerStatus.LastTerminationState.Terminated.Message
+		}
+	}
+	...
+	return ""
+}
+```
+Then we could update the status of backupRepository CR with the message.
+
+### 4. Add Setting for Maintenance Job Timeout
```
Comment (qiuming-best) [inline design/repository-maintenance.md:155]: Not sure we need timeout configuration If the maintenance job timeout, the controller deletes the job which may lead to maintenance breaking halfway.

File: design/repository-maintenance.md [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+
+
+## Detailed Design
+
+### 1. Add Maintenance CLI Command
+
+The CLI command will be added to the Velero CLI.
+
+```shell
+$ velero maintenance --repo-name $repo-name --repo-type $repo-type --backup-storage-location $bsl
```
Comment (Lyndon-Li) [inline design/repository-maintenance.md:46]: velero maintenance -> velero server repo-maintain?
Let's make it as a velero server sub command, so as to indicate users that they should not call this CLI from client

File: design/repository-maintenance.md [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+	log.Debugf("Creating maintenance job: %v", job)
+	defer cancel()
+	if err := m.client.Create(pruneTimeOut, job); err != nil {
+		return errors.Wrap(err, "error to create maintenance job")
+	}
+
+	if err := builder.WaitForJobComplete(pruneTimeOut, m.client, job); err != nil {
+		return errors.Wrap(err, "error to wait for maintenance job complete")
+	}
+	...
```
Comment (Lyndon-Li) [inline design/repository-maintenance.md:126]: Let's get the job result from here, since the job is launched by repoManager, the repoManager should get the result from the job, as the controller doesn't know anything about the job.

File: design/repository-maintenance.md [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+
+### 2. Create Jobs by Repository Manager
+Currently, the backup repository controller will call the repository manager to do the `PruneRepo`, and Kopia or Restic maintenance is then finally called through multiple chain calls.
+
+We will keep using the `PruneRepo` function in the repository manager, but we cut off the multiple chain calls by creating a maintenance job.
+
+```golang
+func (m *manager) PruneRepo(repo *velerov1api.BackupRepository) error {
+	...
+	// writing the maintenance result into default terminationMessagePath
```
Comment (Lyndon-Li) [inline design/repository-maintenance.md:104]: Add some consts for strings used in this function?

File: design/repository-maintenance.md [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+ - Non-blocking maintenance job: it may conflict with updating the same `backuprepositories` CR when parallel maintenance.
+
+ - Maintenance job concurrency control: there is no one suitable mechanism in Kubernetes to control the concurrency of different jobs.
+
+ - Parallel maintenance: Maintaining the same repo by multiple jobs at the same time would have some compatible cases that some providers may not support.
+
+Unfortunately, parallel maintenance is currently not a priority because of the concerns above, improving maintenance efficiency is not the primary focus at this stage.
+
+## High-Level Design
+1. **Add Maintenance CLI Command**: Introduce a new Velero CLI command for repository maintenance.
```
Comment (ywk253100) [inline design/repository-maintenance.md:28]: Let's change it to `Add Maintenance sub Command`. `CLI command` confuses me that seems it runs on client side

File: design/repository-maintenance.md [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+
+### 5. Add Setting for Resource Usage of Maintenance
+Add one configuration for setting the resource limit of maintenance jobs as below:
+```shell
+    velero server --maintenance-job-cpu-request $cpu-request --maintenance-job-mem-request $mem-request --maintenance-job-cpu-limit $cpu-limit --maintenance-job-mem-limit $mem-limit
+```
+Our default value is 0, which means we don't limit the resources.
+
+### 6. Clear up Maintenance Jobs
+Add one configuration for the number of keeping latest failure maintenance jobs as below:
```
[the remark this answers is not in the document]
Comment (ywk253100) [reply design/repository-maintenance.md:170]: Will any succeeded job be kept?

File: design/repository-maintenance.md [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+
+We will keep using the `PruneRepo` function in the repository manager, but we cut off the multiple chain calls by creating a maintenance job.
+
+```golang
+func (m *manager) PruneRepo(repo *velerov1api.BackupRepository) error {
+	...
+	// writing the maintenance result into default terminationMessagePath
+	args := []string {
+	"-c",
+	"velero maintenance --repo-name=" + param.BackupRepo.Spec.VolumeNamespace + " --repo-type=" + param.BackupRepo.Spec.RepositoryType + " --backup-storage-location=" + param.BackupLocation.Name + " > /dev/termination-log",
```
Comment (ywk253100) [inline design/repository-maintenance.md:107]: We should only put the final error message to file `/dev/termination-log` and put the msg in the repo status then.
The whole logs of the maintenance job can be got via the job logs.

File: design/repository-maintenance.md [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+	}
+
+	env := &v1.EnvVar{Name: ProviderFile, Value: "/credentials/cloud"}
+	job := builder.ForJob(param.BackupRepo.Name, param.BackupRepo.Namespace, "velero/velero:maintenance", args).Env(env).
+		VolumeMounts(*builder.ForVolumeMount("cloud-credentials", "/credentials").Result()).
+		ServiceAccount("velero").Resources(&resources).
+		Volumes(*builder.ForVolume("cloud-credentials").SecretSource("cloud-credentials").Result()).Result()
```
Comment (ywk253100) [inline design/repository-maintenance.md:114]: One thing needs to note is that some attributes of the job should come from the Velero pod

File: design/repository-maintenance.md [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+### 6. Clear up Maintenance Jobs
+Add one configuration for the number of keeping latest failure maintenance jobs as below:
+```shell
+    velero server --keep-failed-maintenance-jobs $num
+```
+
+We would clean all completed jobs and keep the latest N failed jobs at the end of each round of reconciliation.
+
+```golang
+func clearFinishedJob(ctx context.Context, c client.Client, retainFailedJobs int) error {
```
Comment (ywk253100) [inline design/repository-maintenance.md:178]: When will the clean-up action be triggered?

File: design/repository-maintenance.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (qiuming-best) [reply design/repository-maintenance.md:170]: Currently, it will delete all succeeded jobs. why should we keep some succeeded jobs? @ywk253100

File: design/repository-maintenance.md (hunk 7 above)
Comment (qiuming-best) [reply design/repository-maintenance.md:107]: yes, you are right

File: design/repository-maintenance.md (hunk 9 above)
Comment (qiuming-best) [reply design/repository-maintenance.md:178]: it will triggered at the end of Reconcile, will it be so frequently?

File: design/repository-maintenance.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (ywk253100) [reply design/repository-maintenance.md:170]: If the first run of the job fails but the second run succeeds, will users need to care about the failed one? Seems not? So we only need to keep the last job?
BTW, the maintenance job is per repository, right? So we need to keep the last x jobs for every repository?

File: design/repository-maintenance.md (hunk 9 above)
Comment (ywk253100) [reply design/repository-maintenance.md:178]: If it is triggered at the end of reconciling, does that mean any change of the repository also triggers the clean-up action? Seems this is unnecessary. 
How about triggering the clean-up action to clean up the corresponding repository's jobs when one job completes?

Comment (qiuming-best) [reply design/repository-maintenance.md:178]: ok, I think it's more reasonable

File: design/repository-maintenance.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (qiuming-best) [reply design/repository-maintenance.md:170]: yes, your suggestion sounds good

File: design/repository-maintenance.md [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+
+The CLI command will be added to the Velero CLI.
+
+```shell
+$ velero maintenance --repo-name $repo-name --repo-type $repo-type --backup-storage-location $bsl
+```
+
+The main maintenance logic would be using the repository provider to do the maintenance.
+```golang
+func (o *Options) Run(f veleroCli.Factory) {
```
[the remark this answers is not in the document]
Comment (qiuming-best) [reply design/repository-maintenance.md:51]: I want the log level inherited from the velero server.

File: design/repository-maintenance.md [hunk 11]
```diff
@@ -0,0 +1,236 @@
...
+	...
+	credentialFileStore, err := credentials.NewNamespacedFileStore(
+		cli,
+		f.Namespace(),
+		"/tmp/credentials",
+		filesystem.NewFileSystem(),
+	)
+	cmd.CheckError(err)
```
[the remark this answers is not in the document]
Comment (qiuming-best) [reply design/repository-maintenance.md:59]: there are two kinds of messages, one is the `fmt.Fprintf` type of message, which would be redirected to termination-log, and another kind of message would be log could be retrieved by job log

File: design/repository-maintenance.md [hunk 12]
```diff
@@ -0,0 +1,236 @@
...
+	err = repoProvider.BoostRepoConnect(context.Background(), para)
+	cmd.CheckError(err)
+	// Prune repo
+	err = repoProvider.PruneRepo(context.Background(), para)
+	cmd.CheckError(err)
+	...
+}
+```
+
+### 2. Create Jobs by Repository Manager
```
[the remark this answers is not in the document]
Comment (qiuming-best) [reply design/repository-maintenance.md:96]: sounds good

File: design/repository-maintenance.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (qiuming-best) [reply design/repository-maintenance.md:170]: ok, I'll keep the latest N jobs for each repo

File: design/repository-maintenance.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (qiuming-best) [reply design/repository-maintenance.md]: I've added the `FileHook` section in the design, which could dump errors into termination-log

File: design/repository-maintenance.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (qiuming-best) [reply design/repository-maintenance.md]: I've added instructions that this command only could by used internally.

If client calls it, it would fail

Comment (qiuming-best) [reply design/repository-maintenance.md]: I've changed into velero server repo-maintenance, the word maintenance which consistent with Kopia maintenance Command

File: design/repository-maintenance.md (hunk 5 above)
Comment (qiuming-best) [reply design/repository-maintenance.md]: modified

File: design/repository-maintenance.md [hunk 13]
```diff
@@ -0,0 +1,236 @@
...
+## High-Level Design
+1. **Add Maintenance CLI Command**: Introduce a new Velero CLI command for repository maintenance.
+
+2. **Create Jobs by Repository Manager**: Modify the backup repository controller to create a maintenance job instead of directly calling the multiple chain calls for Kopia or Restic maintenance.
+
+3. **Update Maintenance Job Result in BackupRepository CR**: Retrieve the result of the maintenance job and update the status of the `BackupRepository` CR accordingly.
+
+4. **Add Setting for Maintenance Job**: Introduce a configuration option to set maintenance jobs, including resource limits (CPU and memory), keeping the latest N failed maintenance jobs.
+
+4. **Clear Up Maintenance Jobs**: Introduce a configuration to determine the number of failed maintenance jobs to retain for debugging purposes.
```
[the remark this answers is not in the document]
Comment (qiuming-best) [reply design/repository-maintenance.md]: ok, now it changes to keep the latest N maintenance jobs for each repository

File: design/repository-maintenance.md [hunk 14]
```diff
@@ -0,0 +1,236 @@
...
+
```
[the remark this answers is not in the document]
Comment (qiuming-best) [reply design/repository-maintenance.md]: @allenxu404 will help to add repo maintenance metrics later

File: design/repository-maintenance.md [hunk 15]
```diff
@@ -0,0 +1,236 @@
...
+func (m *manager) PruneRepo(repo *velerov1api.BackupRepository) error {
+	...
+	// writing the maintenance result into default terminationMessagePath
+	args := []string {
+	"-c",
+	"velero maintenance --repo-name=" + param.BackupRepo.Spec.VolumeNamespace + " --repo-type=" + param.BackupRepo.Spec.RepositoryType + " --backup-storage-location=" + param.BackupLocation.Name + " > /dev/termination-log",
+	}
+
+	env := &v1.EnvVar{Name: ProviderFile, Value: "/credentials/cloud"}
+	job := builder.ForJob(param.BackupRepo.Name, param.BackupRepo.Namespace, "velero/velero:maintenance", args).Env(env).
```
[the remark this answers is not in the document]
Comment (qiuming-best) [reply design/repository-maintenance.md]: ok, I'm not using the builder package, now I've write some utility functions instead

File: design/repository-maintenance.md [hunk 16]
```diff
@@ -0,0 +1,338 @@
...
+      spec:
+        containers:
+        # arguments for repo maintenance job
+        - args:
+          - server
+          - repo-maintenance
+          - --repo-name=nginx-example
+          - --repo-type=kopia
+          - --backup-storage-location=default
+          - --keep-latest-maintenance-jobs=3
```
Comment (ywk253100) [inline design/repository-maintenance.md:194]: With `--keep-latest-maintenance-jobs`, is repo maintenance job responsible to clean up the job history?

File: design/repository-maintenance.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (qiuming-best) [reply design/repository-maintenance.md]: Not sure we need the debug bundle, for its backup or restore related. but maintenance is independent of backup or restore.

Users may collect the maintenance job logs or we extend `velero debug` with a new maintenance flag to collect job logs?

File: design/repository-maintenance.md [hunk 17]
```diff
@@ -0,0 +1,338 @@
...
+	...
+	return ""
+}
+```
+Then we could update the status of backupRepository CR with the message.
+
+### 4. Add Setting for Resource Usage of Maintenance
+Add one configuration for setting the resource limit of maintenance jobs as below:
+```shell
+    velero server --maintenance-job-cpu-request $cpu-request --maintenance-job-mem-request $mem-request --maintenance-job-cpu-limit $cpu-limit --maintenance-job-mem-limit $mem-limit
```
Comment (ywk253100) [inline design/repository-maintenance.md:261]: Will we support setting these options via `velero install`?

File: design/repository-maintenance.md [hunk 18]
```diff
@@ -0,0 +1,338 @@
...
+```
+Our default value is 0, which means we don't limit the resources.
+
+### 5. Clear up Maintenance Jobs
+Add configuration for clear up maintenance jobs:
+	
+- keep-latest-maintenance-jobs: the number of keeping latest maintenance jobs
+
+	```shell
+		velero server --keep-latest-maintenance-jobs $num
```
Comment (ywk253100) [inline design/repository-maintenance.md:271]: Will we support setting this option via `velero install`?

File: design/repository-maintenance.md [hunk 19]
```diff
@@ -0,0 +1,336 @@
...
+	```
+
+We would check and keep the latest N jobs after a new job is finished.
+
+```golang
+// deleteOldMaintenanceJobs deletes old maintenance jobs and keeps the latest N jobs for each repository
+func deleteOldMaintenanceJobs(cli client.Client, repo string, keep int) error {
```
Comment (ywk253100) [inline design/repository-maintenance.md:278]: Kubernetes supports cleaning up jobs automatically with setting `.spec.ttlSecondsAfterFinished`, is it easier for us to leverage this capability? e.g. `.spec.ttlSecondsAfterFinished=n*maintenance-interval`

File: design/repository-maintenance.md (hunk 16 above)
Comment (qiuming-best) [reply design/repository-maintenance.md:194]: It's the `repository manager` response for maintenance jobs including creating, cleaning up jobs.
it's not cleaned by repo maintenance job

File: design/repository-maintenance.md (hunk 17 above)
Comment (qiuming-best) [reply design/repository-maintenance.md:261]: sounds good, I'll add the support

File: design/repository-maintenance.md (hunk 18 above)
Comment (qiuming-best) [reply design/repository-maintenance.md:271]: sounds good, I'll add the support

File: design/repository-maintenance.md (hunk 19 above)
Comment (qiuming-best) [reply design/repository-maintenance.md:278]: it's good way if we choose TTL.

File: design/repository-maintenance.md [hunk 20]
```diff
@@ -0,0 +1,338 @@
...
+
+on the whole, we record two kinds of logs:
+
+- one is the log output of the intermediate maintenance process: this log could be retrieved via the Kubernetes API server, including the error log.
+
+- one is the result of the command which could indicate whether the execution is an error or not: the result could be redirected to a file that the maintenance job itself could read, and the file only contains the error message.
+
+If we want an error message to be recorded into a file, we could use the [hook of logrus](https://pkg.go.dev/github.com/sirupsen/logrus@v1.9.3/hooks/writer#section-readme).
+
+below we defined the `FileHook` struct which implements the `logrus.Hook` interface.
```
[the remark this answers is not in the document]
Comment (qiuming-best) [reply design/repository-maintenance.md:59]: it will record all errors into one file, whether stopping the process is determined by our error handle logic.

File: design/repository-maintenance.md [hunk 21]
```diff
@@ -0,0 +1,338 @@
...
+
+The main maintenance logic would be using the repository provider to do the maintenance.
+
+```golang
+func checkError(err error, log logrus.FieldLogger) {
+	if err != nil {
+		if err != context.Canceled {
+			log.Error("An error occurred: %v\n", err)
+		}
+		os.Exit(1)
```
[the remark this answers is not in the document]
Comment (qiuming-best) [reply design/repository-maintenance.md]: I've changed the check error logic

File: design/repository-maintenance.md [hunk 22]
```diff
@@ -0,0 +1,363 @@
...
+}
+```
+
+### 2. Create Jobs by Repository Manager
+Currently, the backup repository controller will call the repository manager to do the `PruneRepo`, and Kopia or Restic maintenance is then finally called through multiple chain calls.
+
+We will keep using the `PruneRepo` function in the repository manager, but we cut off the multiple chain calls by creating a maintenance job.
+
+The job definition would be like below:
+```yaml
```
Comment (reasonerjt) [inline design/repository-maintenance.md:175]: Could you remove the generated fields and keep the relative information, rather than pasting the output of `kubectl get jobs`....

File: design/repository-maintenance.md (hunk 20 above)
[the remark this answers is not in the document]
Comment (reasonerjt) [reply design/repository-maintenance.md:59]: If the `one file` means `/dev/termination-log` it should contain only the error message that causing the termination of the job.

File: design/repository-maintenance.md [hunk 23]
```diff
@@ -0,0 +1,363 @@
...
+          - /velero
+          # inherit environment variables from the velero deployment
+          env:
+          - name: AZURE_CREDENTIALS_FILE
+            value: /credentials/cloud
+          # inherit image from the velero deployment
+          image: velero/velero:main
+          imagePullPolicy: IfNotPresent
+          name: kopia-maintenance-container
+          resources: {}
```
Comment (reasonerjt) [inline design/repository-maintenance.md:230]: If we deliberately choose not to set any resource limitation here, could you add a comment to explain?
i.e. if no resource limitation is set via velero server, velero will apply the "best effort" approach.

File: design/repository-maintenance.md (hunk 19 above)
Comment (reasonerjt) [reply design/repository-maintenance.md:302]: I think it makes sense.

File: design/repository-maintenance.md [hunk 24]
```diff
@@ -0,0 +1,363 @@
...
+Roughly, the process is as follows:
+1. The backup repository controller will check the BackupRepository request in the queue periodically.
+
+2. If the maintenance period of the repository checked by `runMaintenanceIfDue` in `Reconcile` is due, then the backup repository controller will call the Repository manager to execute `PruneRepo`
+
+3. The `PruneRepo` of the Repository manager will create one maintenance job, the resource usage would follow the setting in the Velero server, and the environment variable, service account, images, etc. would follow the Velero server pod.
+
+4. The maintenance job will execute the Velero maintenance command and write the result into the terminationMessagePath of the related pod.
+
+5. The backup repository controller will wait for the jobs to update the message field and phase in the status of `backuprepositories` CR after finishing maintenance by reading the job terminal message accordingly.
```
Comment (reasonerjt) [inline design/repository-maintenance.md:348]: By reading this item it looks like it's the jobs updating the `backuprepositories` CR, which I believe is not the case.
Could you rephrase to clarify it's the repo controller making the update to the `backuprepositories` CR by reading status of the job.

File: design/repository-maintenance.md (hunk 20 above)
[the remark this answers is not in the document]
Comment (Lyndon-Li) [reply design/repository-maintenance.md:59]: Looks like the sample code mixes below two mechanisms:

- The error message - It should be in `/dev/termination-log `and should include only the errors causing the maintenance fail (in another word, if the maintenance succeeds, `/dev/termination-log ` should be empty)
- The debug log - It is the debug/info/warning generated by this pod and by Kopia and should be there for each job. Particularly, for each maintenance, Kopia generates some internal logs and Velero has the ability to stream the logs (at present, it is printed in Velero server log). So we should generate a general log files like Velero server and node-agent and users should be able to collect the logs by running `velero debug`

File: design/repository-maintenance.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (Lyndon-Li) [reply design/repository-maintenance.md:91]: See this comment https://github.com/vmware-tanzu/velero/pull/7375/files#r1501981162

File: design/repository-maintenance.md (hunk 20 above)
[the remark this answers is not in the document]
Comment (qiuming-best) [reply design/repository-maintenance.md:59]: yes, It only contains error message

File: design/repository-maintenance.md (hunk 22 above)
Comment (qiuming-best) [reply design/repository-maintenance.md:175]: I've removed some redundant fields and added some annotations for some fields.

File: design/repository-maintenance.md (hunk 23 above)
Comment (qiuming-best) [reply design/repository-maintenance.md:223]: Add some comments for resources field

File: design/repository-maintenance.md (hunk 24 above)
Comment (qiuming-best) [reply design/repository-maintenance.md]: rephrased

File: design/repository-maintenance.md (hunk 20 above)
[the remark this answers is not in the document]
Comment (ywk253100) [reply design/repository-maintenance.md:59]: Do we still need the log hook mechanism? Seems doesn't

File: design/repository-maintenance.md [hunk 25]
```diff
@@ -0,0 +1,336 @@
...
+	if err != nil {
+		if err != context.Canceled {
+			log.Errorf("An error occurred: %v", err)
+			return err
+		}
+	}
+	return nil
+}
+
+func (o *Options) Run(f veleroCli.Factory) {
```
Comment (ywk253100) [inline design/repository-maintenance.md:91]: How about removing the implementation detail code and replace with the processing logic if needed? I can see several errors in the code and let's review them in the implementation PR.

Comment (qiuming-best) [reply design/repository-maintenance.md:81]: ok, I've removed some details but kept the main logic in it.

File: design/repository-maintenance.md [hunk 26]
```diff
@@ -0,0 +1,284 @@
...
+	# which contains the result after maintenance
+	message: ""
+	lastMaintenanceTime: ""
+```
+
+Now, the backup repository controller will call the repository manager to create one maintenance job and wait for the job to complete. The Kopia or Restic maintenance multiple chains are called by the job.
+
+### 3. Update the Result of the Maintenance Job into BackupRepository CR
+
+The backup repository controller will update the result of the maintenance job into the backup repository CR.
```
Comment (shubham-pampattiwar) [inline design/repository-maintenance.md:204]: Are we planning on polling on the job status till it gets completed ? like every 5 sec or something ?

File: design/repository-maintenance.md [hunk 27]
```diff
@@ -0,0 +1,284 @@
...
+
+- keep-latest-maintenance-jobs: the number of keeping latest maintenance jobs for each repository.
+
+	```shell
+		velero server --keep-latest-maintenance-jobs $num
+	```
+
+We would specify the `.spec.ttlSecondsAfterFinished` field of a maintenance job by value`Num * MaintenanceFrequency`.
+
+The TTL-after-finished controller assumes that a Job is eligible to be cleaned up TTL seconds after the maintenance job has finished. Automatic clean up please refer to [it](https://kubernetes.io/docs/concepts/workloads/controllers/ttlafterfinished/).
```
Comment (shubham-pampattiwar) [inline design/repository-maintenance.md:242]: +1 on the usage of `ttlafterfinished`

File: design/repository-maintenance.md (hunk 26 above)
Comment (qiuming-best) [reply design/repository-maintenance.md:204]: Currently, the design is a `backup repository controller` that will trigger to create one job and block waiting for it to complete. The polling frequency would be 1 sec.

Keeping the same logic as the previous `backup repository controller` that will trigger to execute one internal maintenance operation and block wait until it finishes.

If using one non-block way, that controller would update one 'backup repositories' cr at the same time making our current work more complicated.

File: design/repository-maintenance.md [hunk 28]
```diff
@@ -0,0 +1,306 @@
...
+
+## Detailed Design
+
+### 1. Add Maintenance sub-command
+
+The CLI command will be added to the Velero CLI, the command is designed for use in a pod of maintenance jobs. 
+
+Our CLI command is designed as follows:
+```shell
+$ velero repo-maintenance --repo-name $repo-name --repo-type $repo-type --backup-storage-location $bsl
```
Comment (draghuram) [inline design/repository-maintenance.md:44]: Does this command trigger maintenance job every time the command is run? If so, wouldn't it be better to let users set a schedule for maintenance jobs? I hope I haven't misunderstood the design.

Comment (qiuming-best) [reply design/repository-maintenance.md:44]: this command calls the Kopia library or Restic to do maintenance and is used in a maintenance job, the command also reads some other configurations from the job such as environment variables and secret, etc. So using this command is less efficient than using Kopia or Restic to handle the repository directly for users.

File: design/repository-maintenance.md (hunk 20 above)
[the remark this answers is not in the document]
Comment (qiuming-best) [reply design/repository-maintenance.md]: Yes, It seems we do not need to use a log hook mechanism.

File: design/repository-maintenance.md (hunk 19 above)
Comment (qiuming-best) [reply design/repository-maintenance.md:241]: We don't use the TTL finally

File: design/repository-maintenance.md (hunk 28 above)
Comment (reasonerjt) [reply design/repository-maintenance.md:44]: The command will be run in a k8s job and the job will be created by velero pod.
We do not expect user to run this command from the client side.

File: design/repository-maintenance.md [hunk 29]
```diff
@@ -0,0 +1,318 @@
...
+      job-name: nginx-example-default-kopia-pqz6c
+    name: nginx-example-default-kopia-pqz6c
+    namespace: velero
+  spec:
+    # Not retry it again
+    backoffLimit: 1
+    # Only have one job one time
+    completions: 1
+    # Not parallel running job
+    parallelism: 1
```
Comment (kaovilai) [inline design/repository-maintenance.md:142]: @shawn-hurley have warned me in the past about using Jobs (from https://github.com/vmware-tanzu/velero/issues/2601 discussion) that this process has to be able to tolerate being started twice.

> Note that even if you specify .spec.parallelism = 1 and .spec.completions = 1 and .spec.template.spec.restartPolicy = "Never", the same program may sometimes be started twice.

[src](https://kubernetes.io/docs/concepts/workloads/controllers/job/#handling-pod-and-container-failures)

Keep that in mind, I haven't reviewed if repo maintenance starting twice "accidentally" would cause an issue or not.

Comment (kaovilai) [reply design/repository-maintenance.md:142]: I assume the starting twice case would happen very close to each other when the jobs object status haven't been updated to Running yet.

Comment (kaovilai) [reply design/repository-maintenance.md:142]: My napkin algorithm would be to have some kind of leader election even if the expected running job is one. Such that if there is more than one, the later one is no-op or stopped.

Comment (qiuming-best) [reply design/repository-maintenance.md:142]: Thanks for reminding me. If the job started twice "accidentally" for the maintenance scenario, it may have a bug but will not cause issues.

As the maintenance job updates the `LastMaintenanceTime` in `backuprepositories` CR only affects the next maintenance time which doesn't decide whether the repository is ready or not.  it may lead to multiple maintenances in one maintenance frequency in this starting twice "accidentally" scenario.

Comment (qiuming-best) [reply design/repository-maintenance.md:142]: Also for the same repository,  only the maintenance job which gets the repository file lock could run maintenance, and it's guaranteed by the repository itself that only one maintenance at one time for each repository

**Review by blackpiglet — APPROVED** [review]

**Review by Lyndon-Li — APPROVED** [review]
